pkg/ec2provider/ec2provider.go

```diff
@@ -0,0 +1,274 @@
+package ec2provider
+
+import (
+	"errors"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/aws-iam-authenticator/pkg/httputil"
+	"sync"
+	"time"
+)
+
+const (
+	// max limit of k8s nodes support
+	max_channel_size = 5000
+	// max number of in flight non batched ec2:DescribeInstances request to flow
+	max_allowed_single_request = 5
+	// default wait interval for the ec2 instance id request which is already in flight
+	default_wait_interval = 50
+	// Making sure the single instance calls waits max till 5 seconds 100* (50 * time.Millisecond)
+	total_iteration_for_wait_interval = 100
+	// Maximum number of instances with which ec2:DescribeInstances call will be made
+	max_instances_batch_size = 100
+	// Maximum time to wait for a new batch call this also depends on if the instance size has
+	// already become 100 then it will not respect this limit
+	max_wait_interval_for_batch = 200
+)
+
+type EC2Provider interface {
+	// Get a node name from instance ID
+	GetPrivateDNSName(string) (string, error)
+	StartEc2DescribeBatchProcessing()
+}
+
+type ec2PrivateDNSCache struct {
+	dnsCache     map[string]string
+	dnsCacheLock sync.RWMutex
+}
+
+type ec2RequestQueue struct {
+	requestQueueMap  map[string]bool
+	requestQueueLock sync.RWMutex
+}
```

Comment:
Hmm, this is not even really a queue, is it? The pattern of a map to book is commonly used when you want a Set. Also, maps in go return the keys in random order, so you won't even get them in the order they were added when you iterate over it.

Can this just be
```go
type ec2Requests struct {
	set  map[string]bool
	lock sync.RWMutex
}
```

Reply:
Good point.

- my Idea was: 
its a logical concept to understand easily as its a request queue where we have inflight requests saved to have lookup.

I will make this change.